src/main/scala/firrtl/transforms/Dedup.scala

```diff
@@ -18,134 +19,301 @@ case class NoDedupAnnotation(target: ModuleName) extends SingleTargetAnnotation[
+            case (old, nuu) => renameMap.rename(old.serialize, nuu.serialize)
+          }
+          newType
+        } else tpe
       }
     }
-    moduleOrder.foreach(n => onModule(moduleMap(n)))
-    (oldModuleMap.toMap, dedupMap.toMap)
+
+    renameMap.setModule(module.name)
+    // Change module internals
+    changeInternals({n => n}, retype, {i => i}, renameModule)(module)
   }
 
-  def run(c: Circuit, noDedups: Seq[String]): (Circuit, RenameMap) = {
-    val moduleOrder = buildModuleOrder(c)
-    val moduleMap = c.modules.map(m => m.name -> m).toMap
+  /**
+    * Deduplicate
+    * @param circuit Circuit
+    * @param noDedups list of modules to not dedup
+    * @param renameMap rename map to populate when deduping
+    * @return Map of original Module name -> Deduped Module
+    */
+  def deduplicate(circuit: Circuit,
+                  noDedups: Set[String],
+                  renameMap: RenameMap): Map[String, DefModule] = {
 
-    val (oldModuleMap, dedupMap) = findDups(moduleOrder, moduleMap, noDedups)
+    // Order of modules, from leaf to top
+    val moduleLinearization = new InstanceGraph(circuit).moduleOrder.map(_.name).reverse
 
-    // Use old module list to preserve ordering
-    val dedupedModules = c.modules.flatMap(m => oldModuleMap.get(m.name).map(_.head))
+    // Maps module name to original module
+    val moduleMap = circuit.modules.map(m => m.name -> m).toMap
 
-    val cname = CircuitName(c.main)
-    val renameMap = RenameMap(dedupMap.map { case (from, to) =>
-      logger.debug(s"[Dedup] $from -> $to")
-      ModuleName(from, cname) -> List(ModuleName(to, cname))
-    })
+    // Maps a module's tag to its deduplicated module
+    val tag2name = mutable.HashMap.empty[String, String]
+
+    // Maps a module's name to its tag
+    val name2tag = mutable.HashMap.empty[String, String]
+
+    // Maps a tag to all matching module names
+    val tag2all =  mutable.HashMap.empty[String, mutable.Set[String]]
+
+    // Build dedupMap
+    moduleLinearization.foreach { moduleName =>
+      // Get original module
+      val originalModule = moduleMap(moduleName)
+
+      // Replace instance references to new deduped modules
+      val dontcare = RenameMap()
+      dontcare.setCircuit("dontcare")
+      //val fixedModule = DedupModules.dedupInstances(originalModule, tag2module, name2tag, name2module, dontcare)
+
+      if(noDedups.contains(originalModule.name)) {
+        // Don't dedup. Set dedup module to be the same as fixed module
+        name2tag(originalModule.name) = originalModule.name
+        tag2name(originalModule.name) = originalModule.name
+        //templateModules += originalModule.name
+      } else { // Try to dedup
+
+        // Build name-agnostic module
+        val agnosticModule = DedupModules.agnostify(originalModule, name2tag, tag2name)
+
+        // Build tag
+        val tag = md5Hash(agnosticModule match {
```

Comment:
Why do we need this secure hash for the tag? Can't we just use `.hashCode` or even just the serialized String of the Module?